pkg/datastore/filedb/codec.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2022 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filedb
+
+import (
+	"encoding/json"
+	"reflect"
+
+	"github.com/pipe-cd/pipecd/pkg/datastore"
+)
+
+// decode checks for the given collection object. If the given collection
+// implements the `datastore.ShardDecoder` interface, its implementation will
+// be used. If not, time order regardless merge logic will be used.
+func decode(col datastore.Collection, e interface{}, parts ...[]byte) error {
+	// In case it's single part contained object, unmarshal it directly.
+	if len(parts) == 1 {
+		return json.Unmarshal(parts[0], e)
+	}
+
+	dcol, ok := col.(datastore.ShardDecoder)
+	if !ok {
+		return merge(e, parts...)
+	}
+	return dcol.Decode(e, parts...)
+}
+
+// merge function unmarshal all parts of the given data to entity e.
+// The data will be merged regardless of its time order, after be merged,
+// the latest UpdatedAt time will be used as the entity UpdatedAt value.
+func merge(e interface{}, parts ...[]byte) error {
+	type model interface {
+		GetUpdatedAt() int64
+	}
+
+	var latest int64
+	for _, p := range parts {
+		if err := json.Unmarshal(p, e); err != nil {
+			return err
+		}
+		me, ok := e.(model)
+		if !ok {
+			return datastore.ErrUnsupported
+		}
+		if latest < me.GetUpdatedAt() {
+			latest = me.GetUpdatedAt()
+		}
+	}
+	// TODO: Find a better way to set value of field UpdatedAt.
+	reflect.ValueOf(e).Elem().FieldByName("UpdatedAt").SetInt(latest)
```

Comment:
We can also define a `updatedAtSetter` interface beside `updatedAtGetter` as above to cast and set.
We just use this reflect magic in case it is not castable.

Reply:
in that case, we have to implement `SetUpdatedAt` for all models struct since the code generated by protobjf does not contain that function, which a bit odd to me, is that okay?
ref: https://github.com/golang/protobuf/issues/65

Reply:
> in that case, we have to implement SetUpdatedAt for all models struct since the code generated by protobjf does not contain that function

Yes, I see. I think that is fine.

Reply:
But you can note that way to your TODO and deal with that later.
We need a workable version first before a perfect one.

Reply:
Addressed by https://github.com/pipe-cd/pipecd/pull/3316/commits/7b05be36e6f55da361d8a4723225c0f1486a03b6 🙏